BLD: add optional dependencies as extras_require in setup.py

JMBurley opened the pull request:
EDIT: This was originally a limited scope PR to bring recommended dependencies into pandas extras_require, but discussion  evolved into 'solve packaging for all optional dependencies' rather than just solve for the `recommended` optional dependencies.  The logic behind this is the same as originally argued for `recommended` dependencies but I updated the text of the post to keep it aligned with the PR contents

---------

Closes #47335. Closes #39164 . Closes #48361 .

Hoping to make this part of the 1.5 milestone.

Optional dependencies should have package mgmt facilitated through pandas.  This will make mgmt of pandas in production docker environments a significantly simpler and allow pandas explicit control over the versions of optional packages it should operate with.

## Actions
Use setup.cfg options as [documented here](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html?highlight=options.extras_require#configuring-setup-using-setup-cfg-files) to make  optional extras for a pandas install.

Once this is done, a user will be able to ensure version alignment by
```
# in requirements.txt or setup.py
pandas[recommended, s3] == 1.22.4  # install pd 1.22.4 and compatible numexpr, bottleneck, s3fs versions
```

The proposed solution is simple to do, simple to maintain, and already done by major libraries in similar situations ([moto](https://github.com/spulec/moto/blob/master/setup.py), [aiobotocore](https://github.com/aio-libs/aiobotocore/blob/master/setup.py#L19)).

Not having this solution wastes user time by forcing them to manually maintain versions of pandas dependencies that are not clearly linked to pandas, save in the brains of users and by manually referring to an install docs page.

## checklist
- [x] closes #47335 , closes #39164 
- [NA] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [NA] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions. (no new functions)
- [x] Added an entry in the latest `doc/source/whatsnew/v1.4.3.rst` file if fixing a bug or adding a new feature. (in `Other` section)

JMBurley committed:
add-recommended-dependencies-as-extras_require-in-setup.cfg

See issue #47335.

https://github.com/pandas-dev/pandas/issues/47335

recommended dependencies should have package mgmt facilitated through pandas.  This will make mgmt of pandas in production docker environments a lot simpler.

JMBurley committed:
Update v1.4.3.rst

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
double backtick code in rst

JMBurley committed:
Merge branch 'add-recommended-dependencies-as-extras_require-in-setup.cfg' of https://github.com/JMBurley/pandas into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
Test failure unrelated to modified code.  `'pandas/tests/plotting/frame/test_frame_color.py::TestDataFrameColor::test_mpl2_color_cycle_str'`

jbrockmendel commented:
> This same test fails in PRs that are approved to be merged so unclear to me if action is required.

Unrelated, don't worry about it.

mroeschke reviewed (changes requested):
As noted in https://github.com/pandas-dev/pandas/issues/39164#issuecomment-1154544902=, I think the correct groupings should be discussed first before creating a pull request as it would be good to get all the optional dependencies in

JMBurley committed:
rebundle under extras_require `recommended`

JMBurley committed:
[options.extras_require] bundled dependencies by featureset

see #39164 for discussion.

JMBurley committed:
note: [options.extras_require] bundled dependencies by featureset

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
Updates made as per discussion in #39164

JMBurley committed:
Update setup.cfg

rollback numba from recommended.  This would necessitate and update to documentation that requires broad agreement from pd-dev-core team that would slow down the overall PR

JMBurley committed:
add adlfs for azure to `access data in cloud`

see #39164 for discussion.  0.6.0 might be an overly restrictive version, but it is compatible

JMBurley committed:
fix extras_require: PyTables is actually `tables` on PyPi

JMBurley committed:
Update setup.cfg

JMBurley committed:
add `all` option to [options.extras_require]

JMBurley committed:
moved changelog to 1.4.4 as 1.4.3 released while this PR was stalled

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Updated to 1.5.0 compliance

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
-
+# optional extras for recommended dependencies
+# see: doc/source/getting_started/install.rst
+recommended =
```
IMO `recommended` is rather vague. I would rename to  `performance` with `bottleneck, numexpr, numba`

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
+    numba>=0.53.0
+    scipy >=1.7.1
+    xarray>=0.19.0
+s3 =
```
IMO `aws` would be more general here

mroeschke replied:
None today, just aligning to how there's a `gcp` section.

Additionally this future proofs just in case there's an aws service in the future that pandas can talk to.

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
+    gcsfs>=2021.05.0
+    pandas-gbq>=0.15.0
+azure =
+    adlfs >=0.6.0
```
As noted in a prior PR, we don't tests `azure` yet so I would not be confident including this as a category

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
+hdf5 =
+    blosc>=1.20.1
+    tables>=3.6.1
+sql-postgressql =
```
Maybe just `postgresql`

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
     pytest>=6.0
     pytest-xdist>=1.31
-
+# optional extras for recommended dependencies
```
Could these be specified in `pyproject.toml` instead?

JMBurley replied:
Yes. `pyproject.toml [project.optional-dependencies]` and and `setup.cfg [options.extras_require]` should be equivalent.  The ecosystem has been changing rapidly over the past few years.

For reference: [PEP-621](https://peps.python.org/pep-0621/#dependencies-optional-dependencies)

If moving this to pyproject.toml, then we should also move over the existing `test = {stuff}` extra.  For cleanliness, I'd suggest merging this PR as-is, and then porting all of this over to `toml` in another PR.

While Python community is moving towards toml as ground-truth for project definitions, I think pandas core  team should discuss before making the change to dependency declaration in `pyproject.toml`

mroeschke replied:
Yeah agreed. We've been trying to prefer `pyproject.toml` over `setup.cfg` but agreed that can be done in another PR

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
+    lxml>=4.6.3
+plot =
+    matplotlib>=3.3.2
+table =
```
Maybe something like `output formatting` might be more general

mroeschke commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -245,6 +245,18 @@ and attributes without holding entire tree in memory (:issue:`45442`).
+
+Optional dependencies version management
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Optional Pandas dependencies can be managed as extras in a requirements/setup file, for example:
```
```suggestion
Optional pandas dependencies can be managed as extras in a requirements/setup file, for example:
```

mroeschke reviewed (commented):
Does this include all the optional dependencies in `actions-38-minimum_versions.yaml`?

JMBurley committed:
simplify sql option names

JMBurley committed:
extras rename: recommended -> performance

JMBurley committed:
remove azure

support is currently unofficial as of 1.5.0

JMBurley commented:
> actions-38-minimum_versions.yaml

- Currently missing `compression` tools but I can bundle them into an option (although I doubt users will want them simultaneously, I don't want to name three individual optional extras for each compression lib).
- Missing `odfpy`, because that is undocumented elsewhere in installs or other sources I looked at in #39164. Will add to the excel options and the install docs.
 - `pyreadstat` for spss.  I will add.

JMBurley committed:
align with actions-38-minimum_versions.yaml

add  specific installs and, where required, missing install documentation for
 - odfpy
 - pyreadstat
 - compression options

JMBurley committed:
Pandas -> pandas in doc

Co-authored-by: Matthew Roeschke <emailformattr@gmail.com>

JMBurley committed:
extras rename: s3 -> aws

see https://github.com/pandas-dev/pandas/pull/47336#discussion_r923930271

JMBurley committed:
extras rename: table -> output_formatting

to be more general in case of future changes

JMBurley committed:
Merge branch 'add-recommended-dependencies-as-extras_require-in-setup.cfg' of https://github.com/JMBurley/pandas into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
bug: `>=` not `=`

simonjayhawkins commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -245,6 +245,18 @@ and attributes without holding entire tree in memory (:issue:`45442`).
+
+  pandas[recommended, s3]>=1.5.0
+
+Available optional dependencies are ``[all, recommended, computation, s3, gcp, azure, excel, parquet, feather,hdf5, sql-postgressql, ssql-mysql, sql-other, html, plot, table]`` (:issue:`39164`)
```
should this list match setup.cfg?

JMBurley replied:
Good catch, yes this should match the names of the extra_options `{name}= {some_lib}>1.0.0`

JMBurley committed:
Apply suggestions from code review

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

JMBurley committed:
align 1.5.0.rst to latest extras_require updates

JMBurley commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -245,6 +245,18 @@ and attributes without holding entire tree in memory (:issue:`45442`).
+.. _whatsnew_150.enhancements.Optional_dependency_management:
+
+Optional dependencies version management
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Done.

JMBurley committed:
1.5.0.rst example updated to use valid extras

JMBurley commented:
Test failure seems unrelated to the PR:

```
FAILED pandas/tests/scalar/timedelta/test_timedelta.py::TestTimedeltas::test_round_sanity[round]
FAILED pandas/tests/scalar/timedelta/test_timedelta.py::TestTimedeltas::test_round_sanity[floor]
FAILED pandas/tests/scalar/timedelta/test_timedelta.py::TestTimedeltas::test_round_sanity[ceil]
FAILED pandas/tests/scalar/timestamp/test_unary_ops.py::TestTimestampUnaryOps::test_round_sanity[round]
FAILED pandas/tests/scalar/timestamp/test_unary_ops.py::TestTimestampUnaryOps::test_round_sanity[ceil]
```
Merging in upstream changes from dev will hopefully fix?

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

jbrockmendel commented:
> Merging in upstream changes from dev will hopefully fix?

No, the test itself has corner cases that sometimes get hit.

simonjayhawkins commented:
> [JMBurley](https://github.com/JMBurley) requested a review from [simonjayhawkins](https://github.com/simonjayhawkins) [28 minutes ago](https://github.com/pandas-dev/pandas/pull/47336#event-7029612344)

PR looks OK although would maybe would need to repeat the usage in doc/source/whatsnew/v1.5.0.rst in doc/source/getting_started/install.rst for future readers. Also keeping the minimum versions in sync is probably going to be a nightmare.

I've not been involved in the issue discussion so will leave to others to give approval to merge.

JMBurley commented:
Thanks @simonjayhawkins!

Quick thoughts:
>  Also keeping the minimum versions in sync is probably going to be a nightmare.
It will take work, but an explicit bit of code that asserts what the minimum versions are will be a lot better than documentation that merely suggests it in a manner disconnected from code (the latter we will always struggle to keep to date, In writing this PR I found packages that were missing from install.rst).

I expect we will have users submitting PRs that keep the minimum versions in `extras_require` correct.   And globally across all pandas users it will save a LOT of time to be able to get versions programmatically rather than reading docs and manually writing versions of pandas sub-packages into requirements/pyproject.toml/setup file

JMBurley committed:
add optional dep mgmt instructions to install.rst

JMBurley commented:
@mroeschke passing tests and read for review.

Thanks for the iterations thus far.

mroeschke commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    numba>=0.53.0
+    numexpr>=2.7.1
+computation =
+    numba>=0.53.0
```
I would remove `numba` from `computation` since it's in performance. `numba`'s use is only for opt-in performance

mroeschke commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    numexpr>=2.7.1
+computation =
+    numba>=0.53.0
+    scipy >=1.7.1
```
```suggestion
    scipy>=1.7.1
```

mroeschke commented on setup.cfg:
```diff
@@ -55,7 +55,85 @@ test =
+ssql-mysql =
+    SQLAlchemy>=1.4.16
+    pymysql>=1.0.2
+sql-other =
```
Yeah although we don't test other other DB flavors besides sqllite, mysql, postgres, in theory other flavors would be supported via sqlalchemy.

I guess it's okay to keep this category, but maybe just calling this label `sql`

mroeschke replied:
Sure fine with `sql-other` then.

mroeschke commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    pyreadstat>=1.1.2
+    python-snappy>=0.6.0
+    pyxlsb>=1.0.8
+    scipy >=1.7.1
```
```suggestion
    scipy>=1.7.1
```

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -270,6 +270,12 @@ For example, :func:`pandas.read_hdf` requires the ``pytables`` package, while
 
+Optional pandas dependencies can be managed as extras (e.g.,``pandas[performance, aws]>=1.5.0``)
+in a requirements.txt, setup or pyproject.toml file.
+Available optional dependencies are ``[all, performance, computation, aws,
```
I think it would be necessary to document which dependency is included in each section, so it would be good to refactor the headers and dependencies below to reflect the new categories (except `all` which should implicitly contain all the dependencies listed here)

JMBurley replied:
Very good idea.  Done.

mroeschke commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    numba>=0.53.0
+    scipy >=1.7.1
+    xarray>=0.19.0
+aws =
```
I think `fsspec` should be included in `aws` and `gcp`

JMBurley replied:
This one of the advantages of the `extras_require` system: we can and should trust packages to manage their own required dependencies.

The `s3fs` package enforces `fsspec` (https://github.com/fsspec/s3fs/blob/main/requirements.txt), ditto `gcsfs` (https://github.com/fsspec/gcsfs/blob/main/requirements.txt) and therefore Pandas should not pin `fsspec`.  It just creates a way for us to break the dependencies tree.

Unless I'm missing a way in which pandas, for aws and gcp, needs `fsspec` functionality independent of  `s3fs`|`gcsfs` sufficient to risk divergent `fsspec` versions from those packages?

mroeschke commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    zstandard>=0.15.2
+# `all` supersets all the above options.
+# Should be maintained as the complete set of pandas optional dependencies
+all =
```
I think this section is still missing some of the dependencies in `ci/deps/actions-38-minimum_verions.yaml`

JMBurley replied:
Which ones?  Not sure what I should modify here.

- fastparquet is excluded as strictly redundant and inferior to pyarrow as per [Tom's comment](https://github.com/pandas-dev/pandas/issues/39164#issuecomment-1178989749)
- fsspec is missing because it is a explicit s3fs, gcsfs dependency that it is better to let them handle
- pytables is a typo in [actions-38-minimum_verions.yaml](https://github.com/pandas-dev/pandas/blob/main/ci/deps/actions-38-minimum_versions.yaml), which should be [tables](https://pypi.org/project/tables/) to match the package name on PyPi.  (`tables` is included in `all`)
- testing-specific libraries are not present as presumably out of scope for general-use pandas install that `all` facilitates.  (I'm happy for anyone who wants to get  under the hood to do `pandas[test,all]`).  Let me know if you disagree.

mroeschke replied:
I think for a "consistent" user experience, `all` should mean all dependencies listed in `doc/source/getting_started/install.rst` even if there are some redundant or unneeded dependencies

mroeschke commented:
> minimum versions in sync is probably going to be a nightmare

We have a precommit hook to ensure the versions are consistent between in `pandas/compat/_optional.py` and `actions-38-minimum_versions.yaml` are in sync. The script is in `scripts/validate_min_versions_in_sync.py`. It would be good to include the dependencies in `setup.cfg` as well

JMBurley committed:
lint scipy optional import

Co-authored-by: Matthew Roeschke <emailformattr@gmail.com>

JMBurley commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    blosc>=1.21.0
+    bottleneck>=1.3.1
+    boto3>=1.22.7
+    brotli>=0.7.0
```
```suggestion
    brotlipy>=0.7.0
```

JMBurley commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    tabulate>=0.8.9
+    xarray>=0.19.0
+    xlrd>=2.0.1
+    xlwt>=1.3.0
+    xlsxwriter>=1.4.3
```
alphabetical
```suggestion
    xlsxwriter>=1.4.3
    xlwt>=1.3.0
```

JMBurley commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    brotlipy>=0.7.0
+    python-snappy>=0.6.0
+    zstandard>=0.15.2
+# `all` supersets all the above options.
```
```suggestion
# `all` supersets all the above options except for `test`.
```

JMBurley commented on setup.cfg:
```diff
@@ -55,6 +55,93 @@ test =
+    python-snappy>=0.6.0
+    zstandard>=0.15.2
+# `all` supersets all the above options.
+# Should be maintained as the complete set of pandas optional dependencies
```
```suggestion
# Should be kept as the complete set of pandas optional dependencies for general use
```

JMBurley committed:
Apply suggestions from code review

JMBurley committed:
detailed extras guidance in install.rst

 - updated numbas to a full recommended dependency with a promotional bullet point like bottleneck and numexpr
- clarified the extra to use for each set of optional dependencies
- made xml an optional extra, because is does have usage outside of read_html.

JMBurley committed:
_optional.py note to keep track of setup.cfg

JMBurley commented:
> > minimum versions in sync is probably going to be a nightmare
> 
> We have a precommit hook to ensure the versions are consistent between in `pandas/compat/_optional.py` and `actions-38-minimum_versions.yaml` are in sync. The script is in `scripts/validate_min_versions_in_sync.py`. It would be good to include the dependencies in `setup.cfg` as well

Good idea, although I'm considering it out of scope for this PR.

There will need to be a larger refactor to read the versions from setup.cfg (which would be potentially wasted effort if they are then moved to pyproject.toml).  And if you can do that, the versions verification should be refactored to be managed directly from that ground truth file and not the hard-pasted VERSIONS dict in `_optional.py`

JMBurley committed:
bug: indent after bullet in install.rst

JMBurley committed:
remove numba from computation extra.

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
@mroeschke ready for rereview: comments addressed, tests passing (except the edge case timedelta in linux that jbrockmendel said was fine)

JMBurley commented:
@mroeschke is this good to merge ahead of the 1.5 milestone at the end of this week?

JMBurley commented:
@mroeschke ready for re-review.

JMBurley commented:
@mroeschke Ready for re-review.

JMBurley commented:
@mroeschke All comments addressed, ready for re-review.

mroeschke commented:
Sorry for the delay here @JMBurley. I will probably have more time to review this after the 1.5 release unless other core team members have time to weigh in.

meeseeksmachine committed:
Backport PR #48197 on branch 1.5.x (DOC: Cleanup 1.5 whatsnew) (#48228)

Backport PR #48197: DOC: Cleanup 1.5 whatsnew

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48232 on branch 1.5.x (CI: Ensure jobs run on 1.5.x branch) (#48235)

Backport PR #48232: CI: Ensure jobs run on 1.5.x branch

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48180 on branch 1.5.x (CI: Switch to large for circleci) (#48251)

Backport PR #48180: CI: Switch to large for circleci

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48245 on branch 1.5.x (CI: Skip test_round_sanity tests due to failures) (#48257)

Backport PR #48245: CI: Skip test_round_sanity tests due to failures

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48240 on branch 1.5.x (Fix mypy erroring on backport branches) (#48259)

Backport PR #48240: Fix mypy erroring on backport branches

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48215 on branch 1.5.x (REGR: properly update DataFrame cache in Series.__setitem__) (#48268)

Backport PR #48215: REGR: properly update DataFrame cache in Series.__setitem__

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48272 on branch 1.5.x (CI: Require s3fs greater than minumum version in builds) (#48276)

Backport PR #48272: CI: Require s3fs greater than minumum version in builds

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48299 on branch 1.5.x (Bump s3fs to 2021.08.00) (#48305)

Backport PR #48299: Bump s3fs to 2021.08.00

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48027 on branch 1.5.x (ENH: Support masks in groupby prod) (#48302)

Backport PR #48027: ENH: Support masks in groupby prod

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #47762 on branch 1.5.x (REGR: preserve reindexed array object (instead of creating new array) for concat with all-NA array) (#48309)

Backport PR #47762: REGR: preserve reindexed array object (instead of creating new array) for concat with all-NA array

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48246 on branch 1.5.x (REGR: iloc not possible for sparse DataFrame) (#48311)

Backport PR #48246: REGR: iloc not possible for sparse DataFrame

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

meeseeksmachine committed:
Backport PR #48314 on branch 1.5.x (DOC: v1.4.4 release date and tidy up release notes) (#48320)

Backport PR #48314: DOC: v1.4.4 release date and tidy up release notes

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

meeseeksmachine committed:
Backport PR #48301 on branch 1.5.x (DEPR: Deprecate positional arguments in pivot) (#48326)

Backport PR #48301: DEPR: Deprecate positional arguments in pivot

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48214 on branch 1.5.x (WEB: Removing links to pdf version of the docs from web and docs) (#48242)

Backport PR #48214: WEB: Removing links to pdf version of the docs from web and docs

meeseeksmachine committed:
Backport PR #48159 on branch 1.5.x (TST: Fix interchange/plotting/groupby test warnings) (#48279)

Backport PR #48159: TST: Fix interchange/plotting/groupby test warnings

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48324 on branch 1.5.x (BUG: Add note in whatsnew for DataFrame.at behavior change) (#48345)

Backport PR #48324: BUG: Add note in whatsnew for DataFrame.at behavior change

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48336 on branch 1.5.x (DOC: Add whatsnew note for #45404) (#48341)

Backport PR #48336: DOC: Add whatsnew note for #45404

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48254 on branch 1.5.x (REF: avoid FutureWarning about using deprecates loc.__setitem__ non-inplace usage) (#48353)

Backport PR #48254: REF: avoid FutureWarning about using deprecates loc.__setitem__ non-inplace usage

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48334 on branch 1.5.x (BUG: read_html(extract_links=all) with no header) (#48350)

Backport PR #48334: BUG: read_html(extract_links=all) with no header

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48265 on branch 1.5.x (CI: Setting up ssh key to upload prod docs) (#48370)

Backport PR #48265: CI: Setting up ssh key to upload prod docs

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48381 on branch 1.5.x (CI: Pin mambaforge image) (#48401)

Backport PR #48381: CI: Pin mambaforge image

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48229 on branch 1.5.x (TST: Test Nullable int floordiv by 0) (#48413)

Backport PR #48229: TST: Test Nullable int floordiv by 0

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48414 on branch 1.5.x (DOC: Add deprecation to is_categorical) (#48418)

Backport PR #48414: DOC: Add deprecation to is_categorical

Co-authored-by: Kevin Sheppard <bashtage@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48264 on branch 1.5.x (BUG: ArrowExtensionArray._from_* accepts pyarrow arrays) (#48422)

* Backport PR #48264: BUG: ArrowExtensionArray._from_* accepts pyarrow arrays

* Add missing import

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48411 on branch 1.5.x (REGR: get_loc for ExtensionEngine not returning bool indexer for na) (#48430)

Backport PR #48411: REGR: get_loc for ExtensionEngine not returning bool indexer for na

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48419 on branch 1.5.x (BUG: ensure to return writable buffer in __dataframe__ interchange for categorical column) (#48441)

Backport PR #48419: BUG: ensure to return writable buffer in __dataframe__ interchange for categorical column

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48444 on branch 1.5.x (CI: Pin ipython version) (#48449)

Backport PR #48444: CI: Pin ipython version

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48380 on branch 1.5.x (DOC: Clarify that objects dtype takes precedence in where) (#48445)

* Backport PR #48380: DOC: Clarify that objects dtype takes precedence in where

* Update generic.py

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

mroeschke committed:
Manual Backport PR #48427 on branch 1.5.x (BLD: Refactor Dockerfile to not install dev enviornment on base) (#48450)

Backport PR #48427: BLD: Refactor Dockerfile to not install dev enviornment on base

meeseeksmachine committed:
Backport PR #48426 on branch 1.5.x (BUG: Column.size should be a method) (#48465)

Backport PR #48426: BUG: Column.size should be a method

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48398 on branch 1.5.x (WARN: Avoid FutureWarnings in tests) (#48420)

* Backport PR #48398: WARN: Avoid FutureWarnings in tests

* Update

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>
Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48416 on branch 1.5.x (REF: ensure to apply suffixes before concat step in merge code) (#48470)

Backport PR #48416: REF: ensure to apply suffixes before concat step in merge code

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48354 on branch 1.5.x (CI: Bump timeout to 180 minutes) (#48474)

Backport PR #48354: CI: Bump timeout to 180 minutes

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48472 on branch 1.5.x (PERF: keep using ObjectEngine for ExtensionArrays for 1.5) (#48486)

Backport PR #48472: PERF: keep using ObjectEngine for ExtensionArrays for 1.5

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48473 on branch 1.5.x (REGR: .describe on unsigned dtypes results in object) (#48501)

Backport PR #48473: REGR: .describe on unsigned dtypes results in object

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

phofl committed:
Backport PR #48443 on branch 1.5.x (BUG: Fix pyarrow groupby tests)  (#48494)

* BUG: Fix pyarrow groupby tests (#48443)

# Conflicts:
#	pandas/tests/extension/test_arrow.py

* CI: Fix failing tests (#48493)

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48490 on branch 1.5.x (CI: Use -j1 for python-dev build to avoid flaky build error) (#48517)

Backport PR #48490: CI: Use -j1 for python-dev build to avoid flaky build error

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke committed:
Manual Backport PR #48428 on branch 1.5.x (BUG/TST: fix a bunch of arraymanager+pyarrow tests) (#48518)

Backport PR #48428: BUG/TST: fix a bunch of arraymanager+pyarrow tests

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48525 on branch 1.5.x (CI: Fix py311 builds different exception message) (#48529)

Backport PR #48525: CI: Fix py311 builds different exception message

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48489 on branch 1.5.x (BUG: fix test_arrow.py tests) (#48532)

Backport PR #48489: BUG: fix test_arrow.py tests

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48543 on branch 1.5.x (DOC: Update footer and include OVH) (#48548)

Backport PR #48543: DOC: Update footer and include OVH

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

mroeschke committed:
Manual Backport PR #48417 on branch 1.5.x (Revert set_index inplace and copy keyword changes) (#48552)

Backport PR #48417: Revert set_index inplace and copy keyword changes

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

meeseeksmachine committed:
Backport PR #48550 on branch 1.5.x (TST: remove 2D tests irrelevant for pyarrow) (#48554)

Backport PR #48550: TST: remove 2D tests irrelevant for pyarrow

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48556 on branch 1.5.x (DOC: Fix docs footer) (#48558)

Backport PR #48556: DOC: Fix docs footer

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48562 on branch 1.5.x (TST: Testing that no warnings are emitted and that inplace fillna produces the correct result (GH48480)) (#48564)

Backport PR #48562: TST: Testing that no warnings are emitted and that inplace fillna produces the correct result (GH48480)

Co-authored-by: RaphSku <45042665+RaphSku@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48563 on branch 1.5.x (DOC: Fix read_sas 1.5 release notes) (#48565)

Backport PR #48563: DOC: Fix read_sas 1.5 release notes

Co-authored-by: Jonas Haag <jonas@lophus.org>

meeseeksmachine committed:
Backport PR #48539 on branch 1.5.x (REGR: groupby doesn't identify null values when sort=False) (#48568)

Backport PR #48539: REGR: groupby doesn't identify null values when sort=False

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48555 on branch 1.5.x (DEPR: Series.astype(np.datetime64)) (#48569)

Backport PR #48555: DEPR: Series.astype(np.datetime64)

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

meeseeksmachine committed:
Backport PR #48557 on branch 1.5.x (WEB: Add new footer to web) (#48571)

Backport PR #48557: WEB: Add new footer to web

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48285 on branch 1.5.x (WEB: Unpin pydata sphinx theme) (#48585)

Backport PR #48285: WEB: Unpin pydata sphinx theme

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48572 on branch 1.5.x (DOC: Fixing styles for the dark theme) (#48584)

Backport PR #48572: DOC: Fixing styles for the dark theme

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48397 on branch 1.5.x (WARN: Remove false positive warning for iloc inplaceness) (#48583)

Backport PR #48397: WARN: Remove false positive warning for iloc inplaceness

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48587 on branch 1.5.x (Fix `series.str.startswith(tuple)`) (#48593)

Backport PR #48587: Fix `series.str.startswith(tuple)`

Co-authored-by: Janosh Riebesell <janosh.riebesell@gmail.com>

meeseeksmachine committed:
Backport PR #48601 on branch 1.5.x (CI: Fix matplolib release issues) (#48617)

Backport PR #48601: CI: Fix matplolib release issues

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48623 on branch 1.5.x (REGR/DOC: Docs left navbar broke) (#48625)

Backport PR #48623: REGR/DOC: Docs left navbar broke

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48619 on branch 1.5.x (REGR: Loc.setitem with enlargement raises for nested data) (#48629)

Backport PR #48619: REGR: Loc.setitem with enlargement raises for nested data

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48627 on branch 1.5.x (DOC: Last changes to release notes for 1.5.0 release) (#48630)

Backport PR #48627: DOC: Last changes to release notes for 1.5.0 release

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

Commit:
RLS: 1.5.0

meeseeksmachine committed:
Backport PR #48642 on branch 1.5.x (DOC: Add release notes for 1.5.1) (#48647)

Backport PR #48642: DOC: Add release notes for 1.5.1

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48639 on branch 1.5.x (CI: Fix directory name for published prod docs) (#48648)

Backport PR #48639: CI: Fix directory name for published prod docs

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48651 on branch 1.5.x (REGR: TextIOWrapper raising an error in read_csv) (#48666)

Backport PR #48651: REGR: TextIOWrapper raising an error in read_csv

Co-authored-by: Torsten Wörtwein <twoertwein@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48599 on branch 1.5.x (DOC: Add deprecation infos to deprecated functions) (#48690)

Backport PR #48599: DOC: Add deprecation infos to deprecated functions

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48620 on branch 1.5.x (REGR: Performance decrease in factorize) (#48710)

Backport PR #48620: REGR: Performance decrease in factorize

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48711 on branch 1.5.x (REGR: Regression in DataFrame.loc when setting df with all True indexer) (#48717)

Backport PR #48711: REGR: Regression in DataFrame.loc when setting df with all True indexer

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48696 on branch 1.5.x (REGR: to_hdf raising AssertionError with boolean index) (#48716)

Backport PR #48696: REGR: to_hdf raising AssertionError with boolean index

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48734 on branch 1.5.x (REGR: Raise on invalid colormap for scatter plot) (#48744)

Backport PR #48734: REGR: Raise on invalid colormap for scatter plot

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48713 on branch 1.5.x (BUG: pivot_table raising Future Warning with datetime column as index) (#48742)

Backport PR #48713: BUG: pivot_table raising Future Warning with datetime column as index

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48703 on branch 1.5.x (Added theme_switcher ) (#48741)

Backport PR #48703: Added theme_switcher

Co-authored-by: Deepak Sirohiwal <38135521+deepaksirohiwal@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48697 on branch 1.5.x (REGR: None converted to NaN when enlarging Series) (#48745)

Backport PR #48697: REGR: None converted to NaN when enlarging Series

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48702 on branch 1.5.x (REGR: dropna affects observed in groupby) (#48750)

Backport PR #48702: REGR: dropna affects observed in groupby

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48782 on branch 1.5.x (REGR: describe raising when result contains NA) (#48793)

Backport PR #48782: REGR: describe raising when result contains NA

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48662 on branch 1.5.x (BUG: Series.getitem not falling back to positional for bool index) (#48799)

Backport PR #48662: BUG: Series.getitem not falling back to positional for bool index

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48751 on branch 1.5.x (WEB: Update link to datapythonista blog url) (#48798)

Backport PR #48751: WEB: Update link to datapythonista blog url

Co-authored-by: Marc Garcia <garcia.marc@gmail.com>

meeseeksmachine committed:
Backport PR #48608 on branch 1.5.x (REGR: assert_index_equal raising with non matching pd.NA) (#48800)

* Backport PR #48608: REGR: assert_index_equal raising with non matching pd.NA

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48785 on branch 1.5.x (BUG: still emitting unnecessary FutureWarning in DataFrame.sort_values with sparse columns) (#48807)

Backport PR #48785: BUG: still emitting unnecessary FutureWarning in DataFrame.sort_values with sparse columns

Co-authored-by: Marco Edward Gorelli <33491632+MarcoGorelli@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48693 on branch 1.5.x (ENH: Make deprecate_nonkeyword_arguments alter function signature) (#48795)

Backport PR #48693: ENH: Make deprecate_nonkeyword_arguments alter function signature

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

rhshadrach committed:
Backport PR #48579 on branch 1.5.x (BUG: Fix calling groupBy(...).apply(func) on an empty dataframe invokes func) (#48817)

BUG: Fix calling groupBy(...).apply(func) on an empty dataframe invokes func  (#48579)

(cherry picked from commit 8b0ad717d1ec54dd40136817a326b41817ffcb86)

Co-authored-by: Dennis Chukwunta <chuksmcdennis@yahoo.com>

rhshadrach committed:
Backport PR #48760 on branch 1.5.x (REGR: groupby.size with axis=1 doesn't return a Series) (#48825)

JMBurley commented:
@mroeschke with pandas 1.5 out, is there now capacity to work on this?

Obviously I'll need to redo some the documentation now that the whatsnew isn't 1.5, and merge conflicts have been created by not moving this in.  

I will resolve both of those if there is confirmation that the work won't go stale again.

mroeschke commented:
Hi @JMBurley so sorry for the delay here. Yes, I should have time later this week to review this PR. I think this is definitely a worthwhile change to include in 2.0

phofl committed:
Backport PR #48820 on branch 1.5.x (BUG: to_datetime(format='...%f') parses nanoseconds) (#48860)

BUG: to_datetime(format='...%f') parses nanoseconds (#48820)

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

phofl committed:
Backport PR #48850 on branch 1.5.x (TYP: Fix typing errors caused by new numpy) (#48859)

meeseeksmachine committed:
Backport PR #48790 on branch 1.5.x (Created icons for dark theme) (#48875)

Backport PR #48790: Created icons for dark theme

Co-authored-by: Lorenzo Vainigli <lorenzovngl@gmail.com>

meeseeksmachine committed:
Backport PR #48805 on branch 1.5.x (Added padding and fixed columns for sponsor logos in mobile view) (#48874)

Backport PR #48805: Added padding and fixed columns for sponsor logos in mobile view

Co-authored-by: Amay Patel <92037532+amay-patel@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48866 on branch 1.5.x (REGR: replace replacing wrong values with inplace and datetime) (#48872)

Backport PR #48866: REGR: replace replacing wrong values with inplace and datetime

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48853 on branch 1.5.x (REGR: Avoid unnecessary warning when setting empty dataframe) (#48873)

Backport PR #48853: REGR: Avoid unnecessary warning when setting empty dataframe

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48833 on branch 1.5.x (BUG: ArrowExtensionArray compared to invalid object not raising) (#48878)

Backport PR #48833: BUG: ArrowExtensionArray compared to invalid object not raising

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

meeseeksmachine committed:
Backport PR #48686 on branch 1.5.x (BUG: to_datetime(tz_mix, utc=True) converts to UTC) (#48882)

Backport PR #48686: BUG: to_datetime(tz_mix, utc=True) converts to UTC

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

phofl committed:
Backport PR #48736 on branch 1.5.x ( BUG: AttributeError: 'function' object has no attribute 'currentframe') (#48887)

phofl committed:
Backport PR #48797 on branch 1.5.x (REGR: fix df.apply with keyword non-zero axis) (#48886)

REGR: fix df.apply with keyword non-zero axis (#48797)

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -245,6 +245,9 @@ Package                                                          Minimum support
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
+pandas recommends the following optional dependencies for performance gains. Can be managed via the
+``performance`` extra when defining the pandas requirement.
```
For each of these sentences specifying which label installs the packages, could we more explicitly say "These dependencies can be specifically installed with `pandas[foo]`"

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -253,6 +256,10 @@ Recommended dependencies
   evaluations. ``bottleneck`` uses specialized cython routines to achieve large speedups. If installed,
   must be Version 1.3.2 or higher.
 
+* `numba <https://github.com/numba/numba>`__: Alternative execution engine for rolling operations.
```
```suggestion
* `numba <https://github.com/numba/numba>`__: Alternative execution engine for reduction operations.
```

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -245,6 +245,9 @@ Package                                                          Minimum support
 Recommended dependencies
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
+pandas recommends the following optional dependencies for performance gains. Can be managed via the
```
I suggest moving this section under `Optional Dependencies` as well.

mroeschke replied:
Happy to have other chime in. My think was more organizationally; these are still optional dependencies so thought it made sense to nest it under the header (they can still be recommended)

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -344,15 +363,19 @@ top-level :func:`~pandas.read_html` function:
 SQL databases
 ^^^^^^^^^^^^^
 
+Can be managed with the ``postgresql, mysql, sql-other`` extras, depending on required sql compatibility.
```
Would be good to add another column to specify which dependency is captured with which label.

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -364,6 +387,8 @@ pymysql                   1.0.2              MySQL engine for sqlalchemy
 Other data sources
 ^^^^^^^^^^^^^^^^^^
 
+Can be managed with the ``hdf5, parquet, feather, spss, excel`` extras, depending on required compatibility.
```
Would be good to add another column to specify which dependency is captured with which label.

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -393,6 +419,8 @@ pyreadstat                1.1.2              SPSS files (.sav) reading
 Access data in the cloud
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
+Can be managed with the ``aws, gcp`` extras, depending on required compatibility.
```
Would be good to add another column to specify which dependency is captured with which label.

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -405,6 +433,8 @@ s3fs                      2021.05.0          Amazon S3 access
 Clipboard
 ^^^^^^^^^
 
+Must be manually managed, depending on operating system.
```
It looks like PyQt4/5 and qtpy could be installed with a `clipboard` label. 

I think I would be `xclip` and `xsel` out of the table and into a `:note:` rst section saying that they need to be installed separately on Linux

mroeschke replied:
> Optional extras should work on all systems or not be included at all

Yeah agreed here.

Mainly for clarity, I was suggesting that `xsel` and `xclip` are separate to a paragraph noting these are not Python packages but OS packages one needs to install for Linux since each table above lists Python packages.

JMBurley replied:
Oh, gotcha.  100% agree on that.  Will change accordingly.

JMBurley committed:
add pandas[foo] throughout to explain the optional extras process

Also `Recommended dependencies` to `Performance dependencies (recommended)`

JMBurley committed:
extend optional_extra`all` to cover tests and redundant packages

JMBurley committed:
add tzdata from pandas 1.5.0

JMBurley commented:
Main updates done.  Wait to see if tests pass.

Then moving comments into doc/source/whatsnew/v1.6.0.rst

JMBurley committed:
Merge remote-tracking branch 'upstream/1.5.x' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
move summary to latest whatsnew doc

Code is now stable and updated.

Attempt a port from doc/source/whatsnew/v1.5.0.rst over to latest document  doc/source/whatsnew/v1.5.1.rst.

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
explicitly pair packages to optional_extra in installs

mroeschke commented on doc/source/whatsnew/v1.5.1.rst:
```diff
@@ -8,7 +8,20 @@ including other versions of pandas.
-.. ---------------------------------------------------------------------------
+.. --------------------------------------------------------------------------
+
+.. _whatsnew_151.enhancements.optional_dependency_management:
```
Unfortunatly, I think this should go in 1.6 as a point releases is more related to bug fixes and regressions from a minor release (1.5)

mroeschke commented on setup.cfg:
```diff
@@ -53,6 +53,103 @@ test =
     hypothesis>=5.5.3
     pytest>=6.0
     pytest-xdist>=1.31
```
Could you add `pytest-asyncio` here? https://github.com/pandas-dev/pandas/pull/48361

JMBurley replied:
Done.  Went for latest version as pin. pytest-asyncio>=0.19.0

mroeschke commented on setup.cfg:
```diff
@@ -53,6 +53,103 @@ test =
+# Also adds the following redundant, superseded packages that are listed as supported:
+#    fastparquet  (by pyarrow https://github.com/pandas-dev/pandas/issues/39164)
+# Should be kept as the complete set of pandas optional dependencies for general use.
+all =
```
Could you write a comment that fsspec is not included with all explicitly since its install with s3fs, gcfs?

mroeschke reviewed (commented):
Looks pretty good. Just noting a few followup tasks (separate PR) that would be good, if you're interested:

1. A CI check that `pip install pandas[foo]` works and install the expected dependencies
2. A pre-commit check that ensure that the min version in setup.cfg are aligned with other areas we specify min version (would mean augmenting `scripts/validate_min_versions_in_sync.py`)

JMBurley commented:
@mroeschke Thanks for the comments.  Will address today.

installs.rst is bugged for unclear reasons so will also try to unpick those errors.  Not clear why it has a problem with the identified lines.

JMBurley committed:
fix sphinx errors in install.rst

extra padding on column + empty return before table

JMBurley committed:
add: pytest-asyncio>=0.19.0

pytest-asyncio>=0.19.0

Closes #48361.

No version guidance available so went for latest.  No-one should be messing around with tests without modern installs.

JMBurley committed:
fsspec note

JMBurley committed:
repin pytest-asyncio>=0.17.0

matches pytest-asyncio>=0.17.0 in ci/deps/actions-38-minimum_versions.yaml

JMBurley committed:
move summary from whatsnew/v1.5.1 to v1.6.0

plus add (:issue:`48361`) resolution note

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
linting

double-backtick ``test``
underline tilde match title length

JMBurley commented:
> Looks pretty good. Just noting a few followup tasks (separate PR) that would be good, if you're interested:
> 
> 1. A CI check that `pip install pandas[foo]` works and install the expected dependencies
> 2. A pre-commit check that ensure that the min version in setup.cfg are aligned with other areas we specify min version (would mean augmenting `scripts/validate_min_versions_in_sync.py`)

Added issues #48942 #48949 to track.  Agree they should be distinct PRs.  Not sure on my availability over next few weeks.

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Add `clipboard` as optional extra

JMBurley commented:
Pending tests passing, I think we are done.  First enhancement for pandas 1.6.0 in the bag!

Will mark for re-review when tests pass.

JMBurley commented:
Ready for review.  Tests pass with a cancellation/timeout on `Python Dev / actions-311-dev (macOS-latest) (pull_request)`.

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

mroeschke commented on doc/source/getting_started/install.rst:
```diff
@@ -253,6 +256,10 @@ Recommended dependencies
   evaluations. ``bottleneck`` uses specialized cython routines to achieve large speedups. If installed,
   must be Version 1.3.2 or higher.
 
+* `numba <https://github.com/numba/numba>`__: alternative execution engine for operations including rolling apply &
```
`including rolling apply & reduction.` -> ```that accept `engine="numba" argument```

mroeschke commented on pandas/core/arrays/sparse/accessor.py:
```diff
@@ -369,7 +369,8 @@ def density(self) -> float:
         tmp = np.mean([column.array.density for _, column in self._parent.items()])
-        return tmp
+        # error: Expression of type "floating" cannot be assigned to return type "float"
+        return tmp  # pyright: ignore[reportGeneralTypeIssues]
```
Is this needed due to this PR?

JMBurley replied:
Nope. I did not do that. Came in via a merge from main.

mroeschke replied:
Could you revert any changed to this file then since this PR should be unrelated?

JMBurley replied:
Done.

mroeschke commented on setup.cfg:
```diff
@@ -53,6 +53,114 @@ test =
+# Also adds the following redundant, superseded packages that are listed as supported:
+#    fastparquet  (by pyarrow https://github.com/pandas-dev/pandas/issues/39164)
+# Other notes:
+#   fsspec is not explicitly listed below because it is a required dependency of s3fs, gcsfs (and we should let them
```
So one argument for listing `fsspec` specifically, and may warrant having a `pandas[filesystem]` category independently, is if users are interacting with a remote filesystem that is not a major cloud-based one (gcsfs, s3fs, etc) e.g. `ftp://`, `file://`, etc

mroeschke replied:
I agree with @TomAugspurger; I also don't think fsspec should be a required dependency since its not all users will use IO functionality in pandas.

TomAugspurger replied:
Just a note that (last I looked) fsspec isn't used for HTTP(s) URLs.

JMBurley replied:
> I also don't think fsspec should be a required dependency since its not all users will use IO functionality in pandas.

We've just added in `pytz` as required in pandas `1.5.0` when not all users will interact with timezones and dates that require that package.  Where do we draw the line for "all users"?  I imagine we don't have exact guidelines...

**I'm happy with whatever packaging decision we can get quorum for.**

In that spirit I've added in a new extra for fsspec; it's an improvement over what we have now and I don't want to delay this PR to try and pitch fsspec as `required`.  

> Just a note that (last I looked) fsspec isn't used for HTTP(s) URLs.

Agreed, anything URL-like goes to fsspec, excluding `http(s)://`.  FWIW I didn't unpack further to understand the http route in pandas (although you certainly could use fsspec alongside requests).

JMBurley committed:
Review comments

 - fss optional extra
 - better numba description

JMBurley committed:
remove unneeded comment

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
Thanks for comments.  All resolved

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

mroeschke commented on doc/source/whatsnew/v1.5.1.rst:
```diff
@@ -8,7 +8,7 @@ including other versions of pandas.
 {{ header }}
 
-.. ---------------------------------------------------------------------------
+.. --------------------------------------------------------------------------
```
Nit: Could you undo this change? (This file will generally receive backports which we do not want to do for this file)

mroeschke commented on setup.cfg:
```diff
@@ -53,6 +53,113 @@ test =
+computation =
+    scipy>=1.7.1
+    xarray>=0.19.0
+fss =
```
I would personally like to see this be called `filesystem` as `fss` isn't immediately obvious to me, but not a blocker to me.

JMBurley replied:
It's a good question.  I went back and forth on `fss` vs `filesystem`.  Ultimately I decided that `filesystem` might suggest to users a broad compatibility with common filesystems (eg. s3) that the optional_extra does not provide, and it was probably least surprising to users to call it `fss`. (and the alternative `fsspec` is worryingly specific in case fsspec were ever deprecated and pandas switched to an alternative filesystem mgmt package in future).

But I don't think either solution is perfect and am very open to alternative suggestions or being argued back to `filesystem`.

JMBurley committed:
fix: leave 1.5.1.rst unchanged by this PR

mroeschke commented on doc/source/whatsnew/v1.5.1.rst:
```diff
@@ -8,7 +8,7 @@ including other versions of pandas.
 {{ header }}
 
-.. ---------------------------------------------------------------------------
+.. -------------------------------------------------------------------------
```
```suggestion
.. ---------------------------------------------------------------------------
```

JMBurley committed:
Update doc/source/whatsnew/v1.5.1.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke reviewed (approved):
Great thanks for your work and patience here, LGTM. Will leave open just in case others want to chime in

JMBurley commented:
> Great thanks for your work and patience here, LGTM. Will leave open just in case others want to chime in

Awesome, glad to have this in.  Optional dependencies are **really** helpful for container mgmt in production DS.

Note: there is now a conflict with main branch where `whatsnew/v1.6.0.rst` has been renamed to `whatsnew/v2.0.0.rst`.  Is pandas 2.0.0 now the next major release instead of pandas 1.6.0?

I'll let someone provide guidance before I alter documentation for the merge.

mroeschke commented:
>  Is pandas 2.0.0 now the next major release instead of pandas 1.6.0?

Correct, 2.0 will be the next major release so a rebase is needed

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
Approved and tests passed!
~~First enhancement for pandas 1.6.0 in the bag!~~
First enhancement for pandas 2.0.0 in the bag!

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley committed:
pyarrow bump 1.0.1 -> 6.0.0

JMBurley commented:
> Sorry this PR hasn't gotten that much other review @JMBurley. I think this will be okay to merge once the conflict is resolved (the pyarrow min version was bumped to 6.0)

Pyarrow version is now bumped & merged in latest from main.  Waiting for tests to pass

JMBurley committed:
Merge branch 'main' into add-recommended-dependencies-as-extras_require-in-setup.cfg

JMBurley commented:
@mroeschke good to merge?

mroeschke commented:
Great work @JMBurley! Thanks for sticking with all the requests. I do think this is really beneficial with all the optional dependencies pandas!

Dr-Irv commented:
@mroeschke I found this part confusing:
```text
Optional pandas dependencies can be managed as optional extras (e.g.,``pandas[performance, aws]>=1.5.0``)
in a requirements.txt, setup, or pyproject.toml file.
Available optional dependencies are ``[all, performance, computation, aws,
gcp, excel, parquet, feather, hdf5, spss, postgresql, mysql, sql-other, html, xml,
plot, output_formatting, compression, test]``
```

I think this should point out that the optional dependencies should ONLY be managed via files, and that you cannot do `conda install pandas[computation]=1.5.1` (for example) to get the dependencies to install.  The way this section is written, it makes it seem that you can do that (or maybe you can with this PR, in which case that should be shown as an installation option).

jorisvandenbossche commented on doc/source/getting_started/install.rst:
```diff
@@ -242,8 +242,11 @@ Package                                                          Minimum support
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+pandas recommends the following optional dependencies for performance gains. These dependencies can be specifically
+installed with ``pandas[performance]`` (i.e. add as optional_extra to the pandas requirement)
```
Echoing @Dr-Irv's comment, also here I find this confusing, since at the top of this document we actually start with mentioning `conda install pandas` before `pip install pandas`. So I would personally remove the `pandas[performance]` here, or at least every time we mention such an extra also point to a dedicated section explaining it?
